ut/s3_put_bucket_versioning_action_test.cc

```diff
@@ -0,0 +1,327 @@
+  std::string bucket_name;
+
+ public:
+  void func_callback_one() { call_count_one += 1; }
+};
+
+TEST_F(S3PutBucketVersioningActionTest, Constructor) {
+  EXPECT_NE(0, action_under_test_ptr->number_of_tasks());
+  EXPECT_TRUE(action_under_test_ptr->put_bucket_version_factory != nullptr);
+}
+
+TEST_F(S3PutBucketVersioningActionTest, ValidateRequest) {
+
+  mock_bucket_version_str =
+      "<VersioningConfiguration"
+      "xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">"
+      "<Status>Enabled</Status>"
+      "</VersioningConfiguration>";
+  call_count_one = 0;
+
+  EXPECT_CALL(*request_mock, has_all_body_content())
+      .Times(AtLeast(1))
+      .WillRepeatedly(Return(true));
+  EXPECT_CALL(*request_mock, get_full_body_content_as_string())
+      .Times(AtLeast(1))
+      .WillRepeatedly(ReturnRef(mock_bucket_version_str));
+  EXPECT_CALL(
+      *(bucket_versioning_body_factory_mock->mock_put_bucket_versioning_body),
+      isOK())
+      .Times(AtLeast(1))
+      .WillRepeatedly(Return(true));
+  EXPECT_CALL(
+      *(bucket_versioning_body_factory_mock->mock_put_bucket_versioning_body),
+      get_versioning_status())
+      .Times(AtLeast(1))
+      .WillRepeatedly(ReturnRef(mock_bucket_versioning_state));
+
+  action_under_test_ptr->clear_tasks();
+  ACTION_TASK_ADD_OBJPTR(action_under_test_ptr,
+                         S3PutBucketVersioningActionTest::func_callback_one,
+                         this);
+  action_under_test_ptr->validate_request();
+  EXPECT_EQ(1, call_count_one);
+}
+
+TEST_F(S3PutBucketVersioningActionTest, ValidateRequestMoreContent) {
+  EXPECT_CALL(*request_mock, has_all_body_content()).Times(1).WillOnce(
+      Return(false));
+  EXPECT_CALL(*request_mock, get_data_length()).Times(1).WillOnce(Return(0));
+  EXPECT_CALL(*request_mock, listen_for_incoming_data(_, _)).Times(1);
+
+  action_under_test_ptr->validate_request();
+}
+
+TEST_F(S3PutBucketVersioningActionTest, ValidateInvalidRequest) {
+
+  mock_bucket_version_str =
+      "<VersioningConfiguration "
+      "xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">"
+      "<Status>Enabled</Status>"
+      "</VersioningConfiguration>";
+
+  EXPECT_CALL(*request_mock, has_all_body_content())
+      .Times(AtLeast(1))
+      .WillRepeatedly(Return(true));
+  EXPECT_CALL(*request_mock, get_full_body_content_as_string())
+      .Times(AtLeast(1))
+      .WillRepeatedly(ReturnRef(mock_bucket_version_str));
+  EXPECT_CALL(
+      *(bucket_versioning_body_factory_mock->mock_put_bucket_versioning_body),
+      isOK())
+      .Times(AtLeast(1))
+      .WillRepeatedly(Return(false));
+  EXPECT_CALL(*request_mock, set_out_header_value(_, _)).Times(AtLeast(1));
+  EXPECT_CALL(*request_mock, send_response(400, _)).Times(AtLeast(1));
+
+  action_under_test_ptr->validate_request();
+  EXPECT_STREQ("MalformedXML",
+               action_under_test_ptr->get_s3_error_code().c_str());
+}
+
+TEST_F(S3PutBucketVersioningActionTest, ValidateRequestXmlTagsWithEnableState) {
```

Comment:
Sorry, where is a tag here?

Reply:
this test case is for validating versioning status of a bucket  which is available in the aws request, by calling below methods.
S3PutBucketVersioningAction::validate_request_xml_tags & S3PutVersioningBody::validate_bucket_xml_versioning_status

Reply:
The function should be named according to the action. This was copied from the PutObjectTag handler which is validating the "tags" fields in the XML. It should be something like `validate_request_xml_versioning_configuration` in this case.

Reply:
done